### PR TITLE
ScreenCordsXY refactor gfx_draw_dashed_line

### PR DESCRIPTION
**Pull request by DrunkBlood**
Lurking on Github for a while I wanted to try my first PR. I looked at the Issues and found some tagged with "Good first issue". So I read the Contributing.md file and tried to follow it as best I can. I selected this #12101 issue and worked on it.

I used 'grep "gfx_draw_dashed_line" *' in the directory src to find all occurrences of the function

For the replacement of calls I am not sure if I should create ScreenLine() inline.
For the include for ScreenLine I am not sure where to place it. I need to include "openrct2/world/Location.hpp" somewhere. I think this needs to be in "openrct2/drawing/Drawing.h" since it other includes delegate to it (Graph.h)

### CODE REVIEW

**Comment by DrunkBlood** [comment]
As expected it fails the check since I did not include ScreenLine yet

File: src/openrct2/drawing/NewDrawing.cpp [hunk 1]
```diff
@@ -212,34 +212,34 @@ void gfx_draw_line(rct_drawpixelinfo* dpi, int32_t x1, int32_t y1, int32_t x2, i
...
 
     const auto drawingEngine = dpi->DrawingEngine;
     if (drawingEngine != nullptr)
     {
         constexpr int32_t precisionFactor = 1000;
 
-        const int32_t dashedLineLength = std::hypot(x2 - x1, y2 - y1);
-        const int32_t lineSegmentCount = dashedLineLength / dashedLineSegmentLength / 2;
+        const int32_t dashedLineLenght = std::hypot(
+            screenLine.GetRight() - screenLine.GetLeft(), screenLine.GetBottom() - screenLine.GetRight());
```
Comment (tupaschoal) [inline src/openrct2/drawing/NewDrawing.cpp:225]: There's a mistake here in the calculation of the length
```suggestion
            screenLine.GetRight() - screenLine.GetLeft(), screenLine.GetBottom() - screenLine.GetTop());
```

File: src/openrct2/drawing/NewDrawing.cpp [hunk 2]
```diff
@@ -212,34 +212,34 @@ void gfx_draw_line(rct_drawpixelinfo* dpi, int32_t x1, int32_t y1, int32_t x2, i
...
+    assert(dashedLineSegmentLenght > 0);
 
     const auto drawingEngine = dpi->DrawingEngine;
     if (drawingEngine != nullptr)
     {
         constexpr int32_t precisionFactor = 1000;
 
-        const int32_t dashedLineLength = std::hypot(x2 - x1, y2 - y1);
-        const int32_t lineSegmentCount = dashedLineLength / dashedLineSegmentLength / 2;
+        const int32_t dashedLineLenght = std::hypot(
```
Comment (tupaschoal) [inline src/openrct2/drawing/NewDrawing.cpp:224]:
```suggestion
        const int32_t dashedLineLength = std::hypot(
```
It's "length" not "lenght"

File: src/openrct2/drawing/NewDrawing.cpp [hunk 3]
```diff
@@ -212,34 +212,34 @@ void gfx_draw_line(rct_drawpixelinfo* dpi, int32_t x1, int32_t y1, int32_t x2, i
 }
 
 void gfx_draw_dashed_line(
-    rct_drawpixelinfo* dpi, const int32_t x1, const int32_t y1, const int32_t x2, const int32_t y2,
-    const int32_t dashedLineSegmentLength, const int32_t colour)
+    rct_drawpixelinfo* dpi, const ScreenLine& screenLine, const int32_t dashedLineSegmentLenght, const int32_t color)
 {
-    assert(dashedLineSegmentLength > 0);
+    assert(dashedLineSegmentLenght > 0);
```
Comment (tupaschoal) [inline src/openrct2/drawing/NewDrawing.cpp:217]:
```suggestion
    assert(dashedLineSegmentLength > 0);
```

File: src/openrct2/drawing/NewDrawing.cpp [hunk 4]
```diff
@@ -212,34 +212,34 @@ void gfx_draw_line(rct_drawpixelinfo* dpi, int32_t x1, int32_t y1, int32_t x2, i
 }
 
 void gfx_draw_dashed_line(
-    rct_drawpixelinfo* dpi, const int32_t x1, const int32_t y1, const int32_t x2, const int32_t y2,
-    const int32_t dashedLineSegmentLength, const int32_t colour)
+    rct_drawpixelinfo* dpi, const ScreenLine& screenLine, const int32_t dashedLineSegmentLenght, const int32_t color)
```
Comment (tupaschoal) [inline src/openrct2/drawing/NewDrawing.cpp:215]:
```suggestion
    rct_drawpixelinfo* dpi, const ScreenLine& screenLine, const int32_t dashedLineSegmentLength, const int32_t color)
```

File: src/openrct2/drawing/Drawing.h [hunk 5]
```diff
@@ -601,7 +601,7 @@ void gfx_draw_pickedup_peep(rct_drawpixelinfo* dpi);
 void gfx_draw_line(rct_drawpixelinfo* dpi, int32_t x1, int32_t y1, int32_t x2, int32_t y2, int32_t colour);
 void gfx_draw_line_software(rct_drawpixelinfo* dpi, int32_t x1, int32_t y1, int32_t x2, int32_t y2, int32_t colour);
 void gfx_draw_dashed_line(
-    rct_drawpixelinfo* dpi, int32_t x1, int32_t y1, int32_t x2, int32_t y2, int32_t dashedLineSegmentLength, int32_t colour);
+    rct_drawpixelinfo* dpi, const ScreenLine& screenLine, const int32_t dashedLineSegmentLenght, const int32_t color)
```
Comment (tupaschoal) [inline src/openrct2/drawing/Drawing.h:604]:
```suggestion
    rct_drawpixelinfo* dpi, const ScreenLine& screenLine, const int32_t dashedLineSegmentLength, const int32_t color)
```

File: src/openrct2-ui/interface/Graph.cpp [hunk 6]
```diff
@@ -248,13 +248,12 @@ static void graph_draw_hovered_value(
     {
         return;
     }
-
-    gfx_draw_dashed_line(dpi, info.coords.x, chartFrame.GetTop(), info.coords.x, info.coords.y, DEFAULT_DASHED_LENGTH, 0);
-    gfx_draw_dashed_line(dpi, chartFrame.GetLeft() - 10, info.coords.y, info.coords.x, info.coords.y, DEFAULT_DASHED_LENGTH, 0);
+    gfx_draw_dashed_line(dpi, ScreenLine(ScreenCoordsXY(info.coords.x, chartFrame.GetTop()), info.coords), DEFAULT_DASHED_LENGTH, 0);
```
Comment (tupaschoal) [inline src/openrct2-ui/interface/Graph.cpp:251]: If it works, use the following short notation
```suggestion
    gfx_draw_dashed_line(dpi, {{ info.coords.x, chartFrame.GetTop() }, info.coords }, DEFAULT_DASHED_LENGTH, 0);
```

File: src/openrct2-ui/interface/Graph.cpp [hunk 7]
```diff
@@ -248,13 +248,12 @@ static void graph_draw_hovered_value(
...
+    gfx_draw_dashed_line(dpi, ScreenLine(ScreenCoordsXY(chartFrame.GetLeft() - 10, info.coords.y), info.coords), DEFAULT_DASHED_LENGTH, 0);
```
Comment (tupaschoal) [inline src/openrct2-ui/interface/Graph.cpp:252]:
```suggestion
    gfx_draw_dashed_line(dpi, { { chartFrame.GetLeft() - 10, info.coords.y }, info.coords }, DEFAULT_DASHED_LENGTH, 0);
```

File: src/openrct2-ui/interface/Graph.cpp [hunk 8]
```diff
@@ -248,13 +248,12 @@ static void graph_draw_hovered_value(
...
 
     if (cursorPosition.y > info.coords.y)
     {
-        gfx_draw_dashed_line(dpi, info.coords.x, info.coords.y, info.coords.x, cursorPosition.y, DEFAULT_DASHED_LENGTH, 0);
+        gfx_draw_dashed_line(dpi, ScreenLine(info.coords, ScreenCoordsXY(info.coords.x, cursorPosition.y), DEFAULT_DASHED_LENGTH, 0);
```
Comment (tupaschoal) [inline src/openrct2-ui/interface/Graph.cpp:256]:
```suggestion
        gfx_draw_dashed_line(dpi, { info.coords, { info.coords.x, cursorPosition.y }}, DEFAULT_DASHED_LENGTH, 0);
```

File: src/openrct2/drawing/NewDrawing.cpp [hunk 9]
```diff
@@ -212,34 +212,34 @@ void gfx_draw_line(rct_drawpixelinfo* dpi, int32_t x1, int32_t y1, int32_t x2, i
...
     const auto drawingEngine = dpi->DrawingEngine;
     if (drawingEngine != nullptr)
     {
         constexpr int32_t precisionFactor = 1000;
 
-        const int32_t dashedLineLength = std::hypot(x2 - x1, y2 - y1);
-        const int32_t lineSegmentCount = dashedLineLength / dashedLineSegmentLength / 2;
+        const int32_t dashedLineLenght = std::hypot(
+            screenLine.GetRight() - screenLine.GetLeft(), screenLine.GetBottom() - screenLine.GetRight());
+        const int32_t lineSegmentCount = dashedLineLenght / dashedLineSegmentLenght / 2;
```
Comment (tupaschoal) [inline src/openrct2/drawing/NewDrawing.cpp:226]:
```suggestion
        const int32_t lineSegmentCount = dashedLineLength / dashedLineSegmentLength / 2;
```
It's "length" not "lenght"

**Comment by tupaschoal** [comment]
Note also that it failed because of the code formatting, check the diff of how it should look here:
https://github.com/OpenRCT2/OpenRCT2/pull/12116/checks?check_run_id=833844377

Note that we do this by using `clang-format`

File: src/openrct2/drawing/Drawing.h [hunk 10]
```diff
@@ -601,7 +601,7 @@ void gfx_draw_pickedup_peep(rct_drawpixelinfo* dpi);
 void gfx_draw_line(rct_drawpixelinfo* dpi, int32_t x1, int32_t y1, int32_t x2, int32_t y2, int32_t colour);
 void gfx_draw_line_software(rct_drawpixelinfo* dpi, int32_t x1, int32_t y1, int32_t x2, int32_t y2, int32_t colour);
 void gfx_draw_dashed_line(
-    rct_drawpixelinfo* dpi, int32_t x1, int32_t y1, int32_t x2, int32_t y2, int32_t dashedLineSegmentLength, int32_t colour);
+    rct_drawpixelinfo* dpi, const ScreenLine& screenLine, const int32_t dashedLineSegmentLength, const int32_t color)
```
Comment (tupaschoal) [inline src/openrct2/drawing/Drawing.h]: Go to line 21 of this file and add this afterwards to fix compilation:
```
    struct ScreenLine;
```

This is called a forward declaration, you provide a type to the compiler, even if incomplete, so you can use reference and pointers of it in the header without actually including its definition. It's used to reduce the number of includes in header files and make compilations generally faster when it's altered.

Comment (DrunkBlood) [reply src/openrct2/drawing/Drawing.h]: Is this something I missed learning c? Or is this a feature of cpp?
And should then ScreenCoordsXY be forward declared twice as in line 19 and 21 of Drawing.h?

Comment (DrunkBlood) [reply src/openrct2/drawing/Drawing.h]: Ahh so I just learned it under another name and I learned to use is in only specific ways. Great that I can expand on that.

Comment (tupaschoal) [reply src/openrct2/drawing/Drawing.h]: Nice :)

**Pull request merged**